server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -999,13 +999,9 @@ private IndexingStrategy planIndexingAsPrimary(Index index) throws IOException {
         assert index.origin() == Operation.Origin.PRIMARY : "planing as primary but origin isn't. got " + index.origin();
         final IndexingStrategy plan;
         // resolve an external operation into an internal one which is safe to replay
-        if (canOptimizeAddDocument(index)) {
-            if (mayHaveBeenIndexedBefore(index)) {
-                plan = IndexingStrategy.overrideExistingAsIfNotThere();
-                versionMap.enforceSafeAccess();
-            } else {
-                plan = IndexingStrategy.optimizedAppendOnly(1L);
-            }
+        final boolean canOptimizeAddDocument = canOptimizeAddDocument(index);
```

Comment:
yes, this will mean that double-delivery of an indexing operation with auto-id and optype create will fail with a version conflict, which is similar to an indexing operation without auto-id but with optype create. I agree that we should try to handle internal retries better (in both situations, with the auto-id case perhaps being simpler), but I also do not want to continue being lenient when it comes to overriding newer docs due to retries, in particular if the optype is create (even more so if we associate an index privilege with this).
My plan was to explore something along the lines of your last paragraph as a follow-up, but I do not see this as a blocker to release the given change here.

Reply:
:+1: